enhancements/builds/volume-secrets.md

```diff
@@ -0,0 +1,225 @@
+---
+title: volume-mounted-secret-and-configmap-injection
+authors:
+  - "@bparees"
+reviewers:
+  - @adambkaplan
+  - @smarterclayton
+approvers:
+  - @adambkaplan
+creation-date: 2020-01-08
+last-updated: 2020-01-08
+status: provisional
+---
+
+# Volume Mounted Secret+ConfigMap Injections
+
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+ > 1. Can we make this the default or even only behavior for builds?  Probably not, need to make it opt-in to avoid potentially breaking existing buildconfigs.
+
+## Summary
+
+Today builds support getting source input from configmaps and secrets.  When users utilize this feature, the configmap or secret is volume-mounted into the build pod.  The next steps depend on whether it is an s2i or dockerfile build.
+
+For s2i builds, the generated Dockerfile contains commands to `ADD` the content at a path specified by the user, the assemble script is invoked, and then the injected content is zeroed out prior to committing the image via a `RUN rm` command added to the Dockerfile.
+
+For dockerfile builds, the user is instructed to add appropriate `ADD` and `RUN rm` commands to their dockerfile to inject the content that is available in the build's working directory (along with their application source, where applicable).
+
+There are a few undesirable aspects to this:
+1) In the dockerfile case, the content can still be found in lower layers of the image unless a layer squashing option is 
+selected.
+2) Requires extra work by the user in the Dockerfile, so each Dockerfile must be customized
+
+This enhancement proposes to introduce an option to use buildah's capability to mount a volume at build time.  The content mounted into the build pod would be then mounted into the container processing the Dockerfile, making that content available within the container so Dockerfile commands could reference it.  No explicit `ADD` would be required, and since mounted content is not committed to the resulting image, no `RUN rm` equivalent is required to clean up the injected content.
+
+
+## Motivation
+
+### Goals
+
+* Simplify how users consume secret + configmap content in builds
+* Increase the security of protected content being injected to images
+* Simplify use cases like mounting subscription entitlement credentials into a build so subscription content can be pulled down during a build while not including the credentials in the resulting image.
```

Comment:
Do we still want to cite this given where https://github.com/openshift/enhancements/pull/214 is headed re: CSI and mounting content via admission yada yada

If that proposal manifests, the openshift/builder image presumably just needs to look in the well understood location to seed buildah and the build API / controller sugar this proposal gives us.

Reply:
Granted you can solve entitlements with this but at this point we probably won't.

Reply:
There are other use cases where credentials can otherwise leak, like private Artifactory creds for npm/maven downloads.

Reply:
technically it could still be used for this use case until the entitlement feature lands (depending which of these lands first) and also might be used to "override" the entitlements that are otherwise being injected.

that said, i'll revise the wording to make it more generically about providing a way to inject credentials that are not published w/ the output image.

Reply:
pushed